src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -75,138 +75,229 @@ public static IEnumerable<VisualStudioInstance> GetInstances(string workingDirec
             {
                 var dotnetSdk = GetInstance(basePath);
                 if (dotnetSdk != null)
+                {
                     yield return dotnetSdk;
+                }
             }
         }
 
-        /// <summary>
-        /// This native method call determines the actual location of path, including
-        /// resolving symbolic links.
-        /// </summary>
-        private static string realpath(string path)
+        private static IEnumerable<string> GetDotNetBasePaths(string workingDirectory)
         {
-            IntPtr ptr = NativeMethods.realpath(path, IntPtr.Zero);
-            string result = Marshal.PtrToStringAuto(ptr);
-            NativeMethods.free(ptr);
-            return result;
-        }
+            AddUnmanagedDllResolver();
 
-        private static string FindDotnetFromEnvironmentVariable(string environmentVariable, string exeName)
-        {
-            string dotnet_root = Environment.GetEnvironmentVariable(environmentVariable);
-            if (!string.IsNullOrEmpty(dotnet_root))
+            string? bestSDK = GetSdkFromGlobalSettings(workingDirectory);
+            if (!string.IsNullOrEmpty(bestSDK))
+            {
+                yield return bestSDK;
+            }
+                
+            string[] dotnetPaths = GetAllAvailableSDKs();
+            // We want to return the newest SDKs first, however, so iterate over the list in reverse order.
+            // If basePath is disqualified because it was later
+            // than the runtime version, this ensures that RegisterDefaults will return the latest valid
+            // SDK instead of the earliest installed.
+            for (int i = dotnetPaths.Length - 1; i >= 0; i--)
             {
-                string fullPathToDotnetFromRoot = Path.Combine(dotnet_root, exeName);
-                if (File.Exists(fullPathToDotnetFromRoot))
+                if (dotnetPaths[i] != bestSDK)
                 {
-                    if (!IsWindows)
-                    {
-                        fullPathToDotnetFromRoot = realpath(fullPathToDotnetFromRoot) ?? fullPathToDotnetFromRoot;
-                        return File.Exists(fullPathToDotnetFromRoot) ? Path.GetDirectoryName(fullPathToDotnetFromRoot) : null;
-                    }
-
-                    return dotnet_root;
+                    yield return dotnetPaths[i];
                 }
             }
 
-            return null;
+            RemoveUnmanagedDllResolver();
```

Comment:
nit: Wrap in `finally` to make sure it runs even if `GetAllAvailableSDKs` throws.